### PR TITLE
[SPARK-40202][PYTHON][SQL] Allow a dictionary in SparkSession.config in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR adds the dictionary support in `SparkSession.config`, see also SPARK-40163

### Why are the changes needed?

For API parity w/ Scala side.

### Does this PR introduce _any_ user-facing change?

Yes. After this change, users can pass a dictionary as below:

```python
>>> SparkSession.builder.config(
...     map={"spark.some.config.number": 123, "spark.some.config.float": 0.123})
```

### How was this patch tested?

Manually tested, and doctest was added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen, @viirya @ueshin FYI

I know that the main reason of adding that signature is because of the type stuff that's not an issue in Python side -  i don't feel strongly if you guys are in doubt. I just thought that it's still a bit useful to pass a dictionary + for a feature parity

**Comment by srowen** [comment]
Same, yeah, it won't matter as much for Python, but wouldn't hurt to add a Python overload to support a dict

**Review by srowen — APPROVED** [review]

**Review by viirya — APPROVED** [review]
Looks useful for Python users.

File: python/pyspark/sql/session.py [hunk 1]
```diff
@@ -202,6 +202,7 @@ def config(
             key: Optional[str] = None,
             value: Optional[Any] = None,
             conf: Optional[SparkConf] = None,
+            map: Optional[Dict[str, "OptionalPrimitiveType"]] = None,
```
Comment (ueshin) [inline python/pyspark/sql/session.py:205]: nit: how about using keyword-only argument?

```py
            conf: Optional[SparkConf] = None,
            *,
            map: Optional[Dict[str, "OptionalPrimitiveType"]] = None,
```

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
We also need to add the overload definition:

```py
        @overload
        def config(self, *, map: Optional[Dict[str, "OptionalPrimitiveType"]]) -> "SparkSession.Builder":
            ...
```

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**